hooks/build

```diff
@@ -0,0 +1,449 @@
+                echo "Skipping pulling image ${IMAGE}"
+                echo
+            else
+                echo
+                set -x
+                docker pull ${IMAGE} || true
+                set +x
+                echo
+            fi
+        fi
+        DOCKER_CACHE_DIRECTIVE="${DOCKER_CACHE_DIRECTIVE} --cache-from ${IMAGE}"
+    done
+
+    echo
+    echo "This build uses Docker cache"
+    echo "Cache directive used: ${DOCKER_CACHE_DIRECTIVE}"
+    echo
+elif [[ "${AIRFLOW_CONTAINER_USE_NO_CACHE}" == "true" ]]; then
+    DOCKER_CACHE_DIRECTIVE="--no-cache"
+    echo
+    echo "Skip cache for builds. Everything will be rebuilt from scratch."
+    echo
+    echo "Cache directive used: ${DOCKER_CACHE_DIRECTIVE}"
+    echo
+else
+    DOCKER_CACHE_DIRECTIVE="--cache-from ${AIRFLOW_CI_IMAGE} --cache-from ${AIRFLOW_IMAGE}"
+    echo
+    echo "Use default cache from locally built images."
+    echo
+    echo "Cache directive used: ${DOCKER_CACHE_DIRECTIVE}"
+    echo
+fi
+
+STAT_BIN=stat
+if [[ "${OSTYPE}" == "darwin"* ]]; then
+    STAT_BIN=gstat
+fi
+# Build id identifying the build uniquely
+BUILD_ID=${BUILD_ID:="local"}
+# Branch name for triggered builds
+BRANCH_NAME=${BRANCH_NAME:="master"}
+
+# directory where "deployment" artifacts should be placed
+DEPLOY_DIR=${AIRFLOW_ROOT}/dist/${BRANCH_NAME}/$(date +%Y-%m-%d)/${BUILD_ID}/${PYTHON_VERSION}
+
+mkdir -p "${DEPLOY_DIR}"
+
+FIXED_PERMISSIONS_FILE=${DEPLOY_DIR}/fixed-permissions.txt
+ALL_FILES_FILE=${DEPLOY_DIR}/all-files.txt
+ALL_FILES_IN_CONTEXT_FILE=${DEPLOY_DIR}/all-files-in-context.txt
+
+mkdir -pv ${DEPLOY_DIR}
+
+if  [[ "${AIRFLOW_FIX_PERMISSIONS}" == "all" || "${AIRFLOW_FIX_PERMISSIONS}" == "setup" ]]; then
+    start_step "Fixing permissions for CI builds for ${AIRFLOW_FIX_PERMISSIONS} files"
+    if [[ "${AIRFLOW_FIX_PERMISSIONS}" == "all" ]]; then
+        # Get all files in the context - by building a small alpine based image
+        # then COPY all files (.) from the context and listing the files via find method
+        if [[ "$(docker images -q airflow-context:latest 2> /dev/null)" == "" ]]; then
+            docker build -t airflow-context:latest .  -f Dockerfile-context
+        fi
+        ALL_FILES_TO_FIX=$(docker run airflow-context:latest /bin/sh -c "(cd /context && find .)")
+    elif [[ "${AIRFLOW_FIX_PERMISSIONS}" == "setup" ]]; then
+        ALL_FILES_TO_FIX="\
+            ${AIRFLOW_ROOT}/setup.py \
+            ${AIRFLOW_ROOT}/setup.cfg \
+            ${AIRFLOW_ROOT}/airflow/version.py \
+            ${AIRFLOW_ROOT}/airflow/__init__.py \
+            ${AIRFLOW_ROOT}/airflow/bin/airflow"
+    fi
+    STAT_BIN=stat
+    if [[ "${OSTYPE}" == "darwin"* ]]; then
+        STAT_BIN=gstat
+    fi
+
+    for FILE in ${ALL_FILES_TO_FIX}; do
+        ACCESS_RIGHTS=$("${STAT_BIN}" -c "%A" "${FILE}" || echo "--------")
+        # check if the file is group/other writeable
+        if [[ "${ACCESS_RIGHTS:5:1}" != "-" || "${ACCESS_RIGHTS:8:1}" != "-" ]]; then
```

Comment:
Why do we care about G/O permissions?

Reply:
It's an intricacy of how git checkout works and how docker build checks invalidation of context combined with traditionally different umask on different platforms. 

I plan to write a blog post about it soon as it is not well known problem.

Git does not store read/write permissions for files (it only stores executable permissions). When you run git checkout on a fresh repo during CI build, the files are created using umask configured for that system. This is controlled by https://git-scm.com/docs/git-config/1.6.3.1#git-config-coresharedRepository setting and default setting is 'umask' - this is how git is configured in all CI services I know :).

When we build image on one system and run --cache-from on a different machine, we want to make sure context is the same otherwise 'COPY setup.py' or 'COPY .' layer will be invalidated. Docker uses a lot of information about each file to detect context change. One of the things it checks are permissions of the file - including group/writeable permissions (it used to take into account file modification date as well but it's been fixed years ago)

The problem is that on different Linux systems can have different umask setting. The most popular umask values are 0002 and 0022 (first gives write access for the group, second only read). Here is some reasoning behind it: https://unix.stackexchange.com/questions/315121/why-is-the-default-umask-002-or-022-in-many-unix-systems-seems-insecure-by-defa

So the problem is that when you checkout the same repo/tag on two systems with different umask, your files will be different so your context will be invalidated.

And - glad that you asked ;) - obviously DockerHub and Travis CI have different umask setting. So when you build image on DockerHub and try to rebuild on Travis it will see them as different context and will start it from re-installing all PIP dependencies (setup.py group-writeable bit is the difference that triggers it). 

So before we run `docker build` I make sure to clear write bits for all files - just to make sure context is the same for all machines.

Reply:
Could you summarize this in a two/three line comment here please?

Reply:
Yep. I Will.